bftengine/src/bftengine/PersistentStorageImp.hpp

```diff
@@ -0,0 +1,220 @@
+  void verifySetDescriptorOfLastNewView(const DescriptorOfLastNewView &desc) const;
+  void verifyLastNewViewMsgs(const DescriptorOfLastNewView &desc) const;
+  void verifyDescriptorOfLastExecution(const DescriptorOfLastExecution &desc) const;
+
+  void saveDescriptorOfLastExitFromView(const DescriptorOfLastExitFromView &newDesc);
+  void setDescriptorOfLastExitFromView(const DescriptorOfLastExitFromView &desc, bool init);
+  void initDescriptorOfLastExitFromView();
+  void saveDescriptorOfLastNewView(const DescriptorOfLastNewView &newDesc);
+  void setDescriptorOfLastNewView(const DescriptorOfLastNewView &desc, bool init);
+  void initDescriptorOfLastNewView();
+  void saveDescriptorOfLastExecution(const DescriptorOfLastExecution &newDesc);
+  void setDescriptorOfLastExecution(const DescriptorOfLastExecution &desc, bool init);
+  void initDescriptorOfLastExecution();
+
+  void setVersion() const;
+
+  void setMsgInSeqNumWindow(const SeqNum &seqNum, const SeqNum &parameterId,
+                            MessageBase *msg, const size_t &msgSize) const;
+  void setBooleanInSeqNumWindow(const SeqNum &seqNum, const SeqNum &parameterId, const bool &boolean) const;
+  void setSeqNumDataElement(const SeqNum &index, char *buf, SharedPtrSeqNumWindow seqNumWindow) const;
+  void serializeAndSaveSeqNumWindow(SharedPtrSeqNumWindow seqNumWindow);
+
+  void setCheckDataElement(const SeqNum &index, char *buf, SharedPtrCheckWindow checkWindow) const;
+  void serializeAndSaveCheckWindow(SharedPtrCheckWindow checkWindow);
+
+  MessageBase *readMsgFromDisk(const SeqNum &seqNum, const SeqNum &parameterId, const size_t &msgSize) const;
+  PrePrepareMsg *readPrePrepareMsgFromDisk(const SeqNum &seqNum) const;
+  FullCommitProofMsg *readFullCommitProofMsgFromDisk(const SeqNum &seqNum) const;
+  PrepareFullMsg *readPrepareFullMsgFromDisk(const SeqNum &seqNum) const;
+  CommitFullMsg *readCommitFullMsgFromDisk(const SeqNum &seqNum) const;
+  bool readBooleanFromDisk(const SeqNum &seqNum, const SeqNum &parameterId) const;
+  void readSeqNumDataElementFromDisk(const SeqNum &index, char *buf, SharedPtrSeqNumWindow seqNumWindow);
+
+  void readCheckDataElementFromDisk(const SeqNum &index, char *buf, SharedPtrCheckWindow checkWindow);
+  CheckpointMsg *readCheckpointMsgFromDisk(const SeqNum &seqNum) const;
+  bool readCompletedMarkFromDisk(const SeqNum &seqNum) const;
+
+  void setFetchingStateInternal(const bool &state);
+  void setLastExecutedSeqNumInternal(const SeqNum &seqNum);
+  void setPrimaryLastUsedSeqNumInternal(const SeqNum &seqNum);
+  void setStrictLowerBoundOfSeqNumsInternal(const SeqNum &seqNum);
+  void setLastViewTransferredSeqNumbersInternal(const ViewNum &view);
+  void setLastStableSeqNumInternal(const SeqNum &seqNum, SharedPtrSeqNumWindow seqNumWindow,
+                                   SharedPtrCheckWindow checkWindow);
+
+ private:
+  MetadataStorage *metadataStorage_ = nullptr;
```

Comment:
OK. Will do it in the next PR.

Reply:
If I might interject a personal opinion, IMHO it would be better if we refrained from approving and merging PRs until we can agree they meet our standards of code quality rather than merging code needing cleanup to master and necessitating more PRs in the future to fix it.

I am concerned that merging code that we already know needs cleanup will lead to the accumulation of technical debt.

Reply:
> If I might interject a personal opinion, IMHO it would be better if we refrained from approving and merging PRs until we can agree they meet our standards of code quality rather than merging code needing cleanup to master and necessitating more PRs in the future to fix it.
> 
> I am concerned that merging code that we already know needs cleanup will lead to the accumulation of technical debt.

 95% of the existing concord-bft code doesn't use smart pointers. When you add new functionality to the existing code and want to use smart pointers, it is almost not feasible to do without changing a lot of other code. In many cases, it is just impossible to do without changing APIs. For this reason, I don't recommend to just replace all pointers by smart ones. The decision should be taken for each case separately.